File: Yandex/Practicum/Algorightms_and_Data_Structures/4_sprint/B.Соревнование/solution.py
import sys


def solution(n: int, points: list[str]) -> None:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    hashmap: dict[int, int] = dict()
    hashmap[0] = 0
    cur_sum: int = 0
    max_sub: int = 0
    for idx in range(n):
        if points[idx] == '0':
            cur_sum -= 1
        else:
            cur_sum += 1
        if cur_sum == 0:
            max_sub = max(max_sub, idx - hashmap[cur_sum] + 1)
        if cur_sum in hashmap:
            max_sub = max(max_sub, idx - hashmap[cur_sum])
        else:
            hashmap[cur_sum] = idx
    sys.stdout.write(str(max_sub) + '\n')


if __name__ == '__main__':
    n: int = int(sys.stdin.readline().rstrip())
    points: list[str] = sys.stdin.readline().rstrip().split()
    solution(n, points)

File: Yandex/Practicum/Algorightms_and_Data_Structures/4_sprint/B.Соревнование/solution.cpp
#include <iostream>
#include <unordered_map>
#include <vector>


void solution(int& n, std::vector<int>& points) {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    std::unordered_map<int, int> hashmap;
    hashmap[0] = 0;
    int maxSub = 0;
    int cumSum = 0;
    for (int idx = 0; idx < n; ++idx) {
        if (points[idx]) {
            ++cumSum;
        } else {
            --cumSum;
        }
        if (cumSum == 0) {
            maxSub = std::max(maxSub, idx + 1);
        }
        if (hashmap.count(cumSum)) {
            maxSub = std::max(maxSub, idx - hashmap[cumSum]);
        } else {
            hashmap[cumSum] = idx;
        }
    }
    std::cout << maxSub << "\n";
}


int main() {
    int n;
    std::cin >> n;
    std::vector<int> points;
    for (int i = 0; i < n; ++i) {
        int point;
        std::cin >> point;
        points.push_back(point);
    }
    solution(n, points);
}